### PR TITLE
[WIP] Generalise Fork Choice Algorithms and implement optimised version.

### PR DESCRIPTION
**Pull request by AgeManning**
## Issue Addressed

This addresses #194. 

## Proposed Changes

This PR will create a `ForkChoice` trait which all fork choice implementations will implement. It allows for generic data structures to be built for different algorithms during block and attestor processing. This will be integrated into the `beacon_chain` struct. 

Potentially add the fork choice to the `chain_spec` struct.

### CODE REVIEW

**Comment by paulhauner** [comment]
I'll let you close this one @AgeManning :)

File: eth2/fork_choice/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+/// Fork choice options that are currently implemented.
+pub enum ForkChoiceAlgorithms {
+    /// Chooses the longest chain becomes the head. Not for production.
+    LongestChain,
+    /// A simple and highly inefficient implementation of LMD ghost.
+    SlowLMDGhost,
+    /// An optimised version of LMD-GHOST by Vitalik.
+    OptimmisedLMDGhost,
```
Comment (paulhauner) [inline eth2/fork_choice/src/lib.rs]: Typo `Optimmised`

File: eth2/fork_choice/src/optimised_lmd_ghost.rs [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        // Note: Votes are weighted by min(balance, MAX_DEPOSIT_AMOUNT) //
+        // FORK_CHOICE_BALANCE_INCREMENT
+        // build a hashmap of block_hash to weighted votes
+        let mut latest_votes: HashMap<Hash256, u64> = HashMap::new();
+        // gets the current weighted votes
+        let current_state = self
+            .state_store
+            .get_reader(&state_root)?
+            .ok_or_else(|| ForkChoiceError::MissingBeaconState(*state_root))?
+            .into_beacon_state()
```
Comment (paulhauner) [inline eth2/fork_choice/src/optimised_lmd_ghost.rs]: There's a short-hand method for this, just FYI

https://github.com/sigp/lighthouse/blob/e4f6fe047da5eab3303dae826b6b4dcd86dafcb6/beacon_node/db/src/stores/beacon_state_store.rs#L22

Comment (paulhauner) [reply eth2/fork_choice/src/optimised_lmd_ghost.rs]: I think it was implemented after `lmd-ghost`

Comment (AgeManning) [reply eth2/fork_choice/src/optimised_lmd_ghost.rs]: Sweet, will update

File: eth2/fork_choice/src/lib.rs (hunk 1 above)
Comment (AgeManning) [reply eth2/fork_choice/src/lib.rs]: Yep, will fix

**Comment by AgeManning** [comment]
Merging down to make a more robust and correct version on the new repos.

**Pull request merged**